### ISSUE TITLE
Incorrect dependency recomputation with class variables

### ISSUE THREAD

**Issue by samkhal** [gristlabs/grist-core#653]
I have the following formula:
```
class Global:
  this_rec = rec
return Global
```
Bug: formulas which access the `this_rec` class variable seem to update only if they are the first record.

Minimal example: https://docs.getgrist.com/sDkEugJfovTo/Dependency-Bug
Observed behavior: modifying `Val1` causes updates on only the first row for the column `Val1 > Val2 indirect`
Expected behavior: all three rows should be updated, as happens in the column `Val1 > Val2 direct`

**Comment by alexmojaki** [comment]
I can see what's happening here. A simpler reproduction is just `return [rec]`.

Record objects are not designed to be reused like this, because the way that the record was created determines how it tracks dependencies (using its `Relation` object). `rec` uses `IdentityRelation` meaning that when a cell uses `rec.foo` then it will only be recalculated when `foo` *in the same row* changes. For changes to `foo` to affect other rows (and other tables) then a different relation is needed, which e.g. lookups provide.

One way to see that this is happening is to change the row ID of the record in Table1. For example, if I add a new row and then delete the first one, now the single record has row ID 2. Then changing `Val1` causes the second row of Table2 to update, not the first row.

When a formula returns a record directly, e.g. `return rec` without the wrapping list, then `_clone_with_relation` gets called which solves the problem in that case.

**Comment by samkhal** [comment]
If I understand correctly, the implication is that it is unsafe for formulas to return objects containing Records, with RecordSet being the only other permitted container. It definitely seems buggy that a row id from one table is used within another table.

Though it looks like `return [rec.Val1]` works perfectly fine. Not sure I understand the difference in terms of the dependency graph. Is there a way to look at the dependency graph between columns for debugging/inspection purposes?

`return [table.lookupOne(id=rec.id)]` fails (of course, same thing as `return [rec]`).
`return lambda: [rec]` and `return lambda: rec` fails.
`return lambda: [table.lookupOne(id=rec.id)]` works.

**Comment by dsagal** [comment]
This gets into very confusing territory about dependency tracking.

TL;DR: if you must return record objects, either return them directly (in a column of type `Any` or of type `Reference`), or as a list, with the column set to type `Reference List`.

Here's an attempt to explain in detail.

The key behavior is that when column `A` uses `some_rec.B`, Grist knows (from evaluating the formula, _not_ from parsing it) that column `A` depends on column `B`, so that changes to `B` need to trigger a recalculation of the `A` formula.

When `some_rec` is the special `rec` variable, it's a one-to-one mapping (change to `B` in row 17 triggers a recalculation of `A` also in row 17), and when `some_rec` comes from a reference column or a lookup, this relationship is more complicated (as in the example, where a change to row 1 in one table triggers a recalculation to multiple rows in the other table).

To keep track of these, Grist uses a concept of `Relation`, attached to each record object, to know where it came from.

In the example in the linked doc, let's change `Table1.G` to the formula `rec`, and add in `Table2.X` this formula:
```
r1 = Table1.lookupOne()
r2 = r1.G
[r1, r2, r2.Val1]
```
1. `r1` comes from `Table1.lookupOne()`. It is a record object wth a relation to tell where it came from: from the lookup. Here we have the trivial lookup that always returns the same value, but in general, the value returned by the lookup can change, and the relation is aware of this.
3. `r2 = r1.G` gets a value from column `G` of `Table1`. Which row to get depends on `r1`. That's where the relation comes into play. Column `Table2.X` with this expression knows that it depends on column `Table1.G`, and the specific row each cell depends on is determined by the lookup relation associated with `r1`. (Again, here it so happens that there is only one row in Table1; in general the dependency row is determined by the lookup.)
4. Now, `r2` is itself a record object in the table `Table1` (because column `G` with the formula `rec` returns such a record). So when we call `r2.Val1`, we know that `Table2.X` depends on column `Table1.Val1`. But how to say which row of `Table2.X` depends on which row of `Table1.Val1`? That's a combination of how we got to the cell of `G` (via a lookup) and how that cell of `G` got `rec` (trivial relation).

The last step is the where all the difficulty is. By itself, the `rec` returned by the formula in `G` just knows that it's the same record as the cell it's evaluated in. When `rec.Val1` is used, it allows following the "identity" relation: a change to column `Val1` row 17 in Table1 should trigger a recalculation of any cell in row 17 of Table1 that mentions `rec.Val1`. But when `rec` is used elsewhere (as above, via `r2` which is `r1.G` which comes from a lookup), it's a composition of relations: a change to column `Val1` row 17 in Table1 follows the identity relation and then the lookup relation, so that we find out which row(s) of Table2 need to get recalculated.

The only place we figure out where this composed relation needs to be created is actually in this bit of code: https://github.com/gristlabs/grist-core/blob/a18007ae7a3841275507add48e73a8e91a982fbd/sandbox/grist/table.py#L698

When we evaluate `r1.G` and get a record object, we know something interesting is going on -- namely that we got a record object, we need to keep track of dependencies, and that requires a composed relation. At that point we create that composed relation.

But now if you change `G` to the formula `[rec]`, and `r1.G` returns a list (rather than a record object), there is no way to tell that this list has something special inside it. So we don't do the special relation composition that's needed to track dependencies. Same if a record object is wrapped into something else, like a `class` in the original example.

When you need to return a list, dependency tracking is correct if you set the column type of `G` to `Reference List`. Then Grist knows enough about `G` values containing records, and constructs the right relations. (In that case, you can't use `r1.G[0]`, but you can use `r1.G.get_one()`, which is equivalent.)

I don't know whether to consider it a bug. Dependency tracking only works when used as intended. It's possible to do really weird things in Python, and capturing them all is impossible. One option is to throw an error when we detect an unexpected usage (it might be possible generally, though I am not 100% sure), but I don't know if that's really better.

**Comment by samkhal** [comment]
Thanks for the detailed explanation!

I agree that capturing all possible ways in which a record could be returned is infeasible. But rather than catching the dependency when that `return` happens, couldn't it be captured later when we perform the `rec_of_unknown_origin.Val1` field access? 

To make up some syntax, my understanding is that:
- For the formula `rec_other = Table1.lookupOne(); rec_other.G` in `Table2.X`, the resulting dependency is `Lookup(Table2, Table1) ∘ Identity(Table1)`. The two relations get composed.
- For the formula `opaque = Table1.lookupOne(); opaque.get_rec().G`, we have `Union(Lookup(Table2, Table1), Identity(Table1))`. Using `Identity(Table1)` within `Table2` gives us the weird behavior where a change to a record in Table1 with id=2 causes an update on the Table2 record with id=2.

Is my understanding approximately correct? If so:
- Suggestion 1: `IdentityRelation` knows the table it's relevant to; if we are trying to use it to update Table2, I think throwing an error would make more sense rather than using an inappropriate index.
- Suggestion 2: Why not perform the composition of relations after-the-fact when evaluating the cell in Table2? If we see that a cell has both relations of `Lookup(Table2, Table1)` and `Identity(Table1)`, that should be enough to compose them. This will overestimate dependencies; if we perform two lookups on the same table and then access a field of a record, we don't know which record it came from. But we can still produce a set of relations `Union(LookupA(Table2,Table1) ∘ Identity(Table1), LookupB(Table2,Table1) ∘ Identity(Table1))`, which should ensure a complete dependency graph. Or to put it another way, since there's a limited number of "supported" ways to access data in other places (`rec`, `Table.lookup*`, `Table.all`), we don't necessarily need to track the full data flow if we can pessimistically compose relations.

I think this should count as a bug; it's unexpected and undocumented behavior. If this really is working as intended, I think it warrants mention in the docs.

**Comment by dsagal** [comment]
Do you have any ideas how to fix it when the lookup is in the same table? I.e. `Lookup(Table1, Table1)` using your notation?

**Comment by samkhal** [comment]
I would imagine it'd work the same way. To generalize: if a formula uses `rec` directly, you add `Identity(table)` to the union of relations as you do now. If a formula accesses a record via any other mechanism, instead of adding `Identity(TableX)`, you compose it with all other compatible lookup relations in the same cell (`Lookup(TableY, TableX) ∘ Identity(TableX)`).

So if you have a lookup in the current table followed by a `other_rec.A`, where `other_rec` is also in the same table, you'd end up with `Lookup(Table1, Table1) ∘ Identity(Table1)`.
